### DISCUSSION TITLE
A small prototype of an UI for the xtc platform

### DISCUSSION THREAD

**Discussion by azzazzel** [xtclang/xvm#89] [Show and tell]
I've been using the platform as a learning playground for some time now. I was curious if I could learn xtclang (in my limited spare time) well enough to build something more than "hello worlds" and counters. It was a nice experience touching almost every part of the platform (DB, model, managers, web, rest services), and I learned a lot. 

After a few nights and weekends, [this](https://www.youtube.com/watch?v=6SpgNVWAaag) is what I got. I'm sharing it with you in case it helps you in any way with the UI you are building. 

#### Look and feel 
I just used an online brand generator. I gave it the phrase "Ecstasy Cloud", and it proposed several designs (logo, colors, fonts, ...). I'm not a designer myself, so please don't judge it from that perspective.

#### Front-end framework
I used [Quasar](https://quasar.dev), which is a component framework based on Vue.js. I'm far from being a front-end expert, but that framework has proven useful in developing not-too-ugly frontends with minimum effort.

#### The code
The code ... is ... a mess. Seriously! It started as a learning playground and still looks like that. But if you still want to have a look at it, it's [here](https://github.com/azzazzel/xtc_platform/tree/quasar_gui). It should be possible to build it and run it locally. In order to work around [#platform/1](https://github.com/xtclang/platform/issues/1), I had to increase the memory for `xec` to 2GB 

#### What's next?
I'd love to try rebuilding the platform in a more consistent and DDD way, but I don't know if I can find the time to do that. 
Anyway, if you find any of this useful, I'm happy to get on a call and provide more details.

**Comment by ggleyzer** [comment]
Thank you! Will take a look

**Reply by ggleyzer** [reply]
I'm struggling to build it. Can we please schedule a call?

Update: after a number of "npm install" commands I managed to build it. It's a black magic tough...)

**Reply by ggleyzer** [reply]
What you did is amazing!! How do you run it "locally"; where to set "process.env.DEV"?

**Reply by azzazzel** [reply]
I'm glad you managed to build it. I was about to make a Dockerfile to make it easier for you to run it. 

So the `process.env.DEV` indicates Quasar (the front-end app) is in dev mode. You can run `yarn --ignore-engines quasar dev` _(or just `quasar dev` if you have the CLI installed)_ from `platofrmUI/gui` folder. This will start just the UI without the backend in dev mode. This is useful for making changes to the UI immediately visible in the browser. But as the backend is not running, you can't make the ajax calls. What the gradle build runs is `yarn --ignore-engines quasar build` which generates a "production ready" version of the UI (minimized, compressed, ...)

**Reply by azzazzel** [reply]
Anyways, I'm still open to having a call if you think it'll be helpful. Feel free to ping me on mail/chat/whatsapp/... I'll send you an email with my contact details.

**Reply by lagergren** [reply]
I would love to have a Dockerfile to use this myself. We should have Dockerfiles for as much as we can, which also gives us great dev container integration.

**Reply by lagergren** [reply]
To elaborate at the request of @gene: just like @azzazzel suggested, it would be great to have a Dockerfile. I run stuff on several platforms, different machines, and in various dev containers. 

The installation instructions are here: https://github.com/azzazzel/xtc_platform/tree/quasar_gui, and contain seven steps that have to be performed in every scratch space and every dev container and on every machine I want to run this on.  I also very much don't like doing sudo changes to my machines for the loopback binding, or any other privileged changes to my base systems, like going into /etc/hosts. A couple of these projects gets messy, and I have a Vite/Vue app somewhere else, for example, that also wanted me to manipulate system state with sudo. 

So I think what @azzazzel suggests is to add a Dockerfile, where the sudo changes, the etc/hosts changes etc are made with RUN commands in the Dockerfile to the container, and the container user can also get the key generated and copied to an appropriate place in the container fs.  The ENTRYPOINT is likely just [ gradle run ], and we can use a base image where java/keytool/gradle are already installed. Then it would be a simple docker run command, which a port forwarding of 8080 to the container host.

**Comment by lagergren** [comment]
Let me try to see if I can use my docker templates to recreate this quickly. I need a short break from minimising the Gradle xtc plugin pull request.

**Comment by azzazzel** [comment]
I actually have a Dockerfile that builds and runs the platform. But I'm
facing some issues with the self signed certificate and it's not accessible
from outside the container. I'll try to fix that tonight and commit the
Dockerfile to the repo.

pt., 11 sie 2023, 14:56 użytkownik Marcus Lagergren <
***@***.***> napisał:

**Reply by lagergren** [reply]
Thanks! I have trouble adding the loopback interface to my docker container. Not sure if I have to tweak the bridge to ping that, but I basically had everything else building, so I'll pause this for now and attack gradle included build names

**Reply by thegridman** [reply]
Do you need to do all that messing about with loopback if you are in a container. Isn't that just to map host names to the loopback address which there might be a better way to do inside a container.

**Reply by azzazzel** [reply]
I'm giving up as well. It seems the platform binds itself to `127.0.0.10`, which is unreachable from outside the Docker container. I can access the platform from within the container but not from outside. I think what's needed is to be able to tell the platform which IP address it should use on start so one can pass the container IP. But I couldn't figure out how to do that.

That said, I committed my [Dockerfile](https://github.com/azzazzel/xtc_platform/blob/quasar_gui/Dockerfile) to the clone if it helps anyone.

**Reply by ggleyzer** [reply]
You can trivially tell the platform what address to use by creating "cfg.json" file, that kernel.x reads:

                String hostName  = config.getOrDefault("hostName",    "xtc-platform.xqiz.it").as(String);
                String bindAddr  = config.getOrDefault("bindAddress", "xtc-platform.xqiz.it").as(String);
                UInt16 httpPort  = config.getOrDefault("httpPort",     8080).as(IntLiteral).toUInt16();
                UInt16 httpsPort = config.getOrDefault("httpsPort",    8090).as(IntLiteral).toUInt16();

As for visibility from the outside, I'm looking into [ngrok](https://ngrok.com/) for that

**Comment by cpurdy** [comment]
What day/time works for everyone to jump on a call? I assume @ggleyzer @lagergren and Mark Schmidt should be present, and I'd love to join! Would Monday morning (NY time) work?

**Comment by thegridman** [comment]
Having done a lot of running in containers work in my day job, I wouldn't mind joining too

**Comment by thegridman** [comment]
I was able to get the platform running in a container and connect my browser to it. You don't need to mess about with loopback, just use the `-h xtc-platform.xqiz.it` option for `docker run` to tell the container its hostname. I have not actually tried uploading anything yet though

**Comment by ggleyzer** [comment]
I pushed the new UI (with minor changes) to [the platform repo](https://github.com/xtclang/platform).

Thank you, @azzazzel!
